stdlib/Printf/src/Printf.jl

```diff
@@ -2,1297 +2,770 @@
-    else
-        push!(blk.args, :(unsafe_write(out, pointer(buf), len)))
-        push!(blk.args, :(while pt >= (len+=1) print(out,'0') end))
-        '#' in flags && push!(blk.args, :(print(out, '.')))
-    end
-    # print space padding
-    if padding !== nothing && '-' in flags
-        push!(blk.args, pad(width-1, padding, ' '))
-    end
-    # return arg, expr
-    :(($x)::Real), ex
-end
+For arbitrary precision numerics, you might extend the method like:
 
-function gen_e(flags::String, width::Int, precision::Int, c::Char, inside_g::Bool=false)
-    # print float in scientific form:
-    #  [e]: use 'e' to introduce exponent
-    #  [E]: use 'E' to introduce exponent
-    #
-    # flags:
-    #  (#): always print a decimal point
-    #  (0): pad left with zeros
-    #  (-): left justify
-    #  ( ): precede non-negative values with " "
-    #  (+): precede non-negative values with "+"
-    #
-    x, ex, blk = if inside_g
-        @gensym x
-        blk = Expr(:block)
-        x, blk, blk
-    else
-        special_handler(flags,width)
-    end
-    # interpret the number
-    if precision < 0; precision = 6; end
-    ndigits = min(precision+1,length(Grisu.getbuf())-1)
-    push!(blk.args, :((do_out, args) = ini_dec(out,$x,$ndigits, $flags, $width, $precision, $c, buf)))
-    push!(blk.args, :(digits = buf))
-    ifblk = Expr(:if, :do_out, Expr(:block))
-    push!(blk.args, ifblk)
-    blk = ifblk.args[2]
-    push!(blk.args, :((len, pt, neg) = args))
-    push!(blk.args, :(exp = pt-1))
-    expmark = isuppercase(c) ? "E" : "e"
-    if precision==0 && '#' in flags
-        expmark = string(".",expmark)
-    end
-    # calculate padding
-    padding = nothing
-    width -= precision+length(expmark)+(precision>0)+4
-    # 4 = leading + expsign + 2 exp digits
-    if '+' in flags || ' ' in flags
-        width -= 1 # for the sign indicator
-        if width > 0
-            padding = quote
-                padn=$width
-                if (exp<=-100)|(100<=exp)
-                    if isa($x,SmallNumber)
-                        padn -= 1
-                    else
-                        padn -= Base.ndigits0z(exp) - 2
+```julia
+Printf.tofloat(x::MyArbitraryPrecisionType) = BigFloat(x)
+```
+"""
+tofloat(x) = Float64(x)
+tofloat(x::Base.IEEEFloat) = x
+tofloat(x::BigFloat) = x
+
+@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Floats}
+    leftalign, plus, space, zero, hash, width, prec =
+        spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision
+    x = tofloat(arg)
+    if x isa BigFloat && isfinite(x)
+        ptr = pointer(buf, pos)
+        newpos = @ccall "libmpfr".mpfr_snprintf(ptr::Ptr{UInt8}, (length(buf) - pos + 1)::Csize_t, string(spec; modifier="R")::Ptr{UInt8}; arg::Ref{BigFloat})::Cint
+        newpos > 0 || error("invalid printf formatting for BigFloat")
+        return pos + newpos
+    elseif x isa BigFloat
+        x = Float64(x)
+    end
```

Comment:
I may be wrong, but I think @vtjnash was thinking something along these lines:

```suggestion
tofloat(x) = Float64(x)
tofloat(x::Base.IEEEFloat) = x
tofloat(x::BigFloat) = x

@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Floats}
    fmt(buf, pos, tofloat(arg), spec::Spec{T}) where {T <: Floats}
end

@inline function fmt(buf, pos, x::BigFloat, spec::Spec{T}) where {T <: Floats}
    if !isfinite(x)
         return fmt(buf, pos, Float64(x), spec)
    end
    leftalign, plus, space, zero, hash, width, prec =
    spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision
    ptr = pointer(buf, pos)
    newpos = @ccall "libmpfr".mpfr_snprintf(ptr::Ptr{UInt8}, (length(buf) - pos + 1)::Csize_t, string(spec; modifier="R")::Ptr{UInt8}; arg::Ref{BigFloat})::Cint
    newpos > 0 || error("invalid printf formatting for BigFloat")
    return pos + newpos
end
@inline function fmt(buf, pos, x::Base.IEEEFloat, spec::Spec{T}) where {T <: Floats}
    leftalign, plus, space, zero, hash, width, prec =
    spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision

```

Reply:
Meh, I'm not a huge fan of splitting methods just for dispatch when a trivial `isa` allows the same compilation benefits.